### ISSUE TITLE
[consistent-type-imports] conflict with importsNotUsedAsValues compile option.

### ISSUE THREAD

**Issue by CatsMiaow** [typescript-eslint/typescript-eslint#4268]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/main/docs/linting/TROUBLESHOOTING.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/consistent-type-imports": "error"
  }
}
```
tsconfig.json
```json
{
  "compilerOptions": {
    "importsNotUsedAsValues": "error",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true
  }
}
```

```TS
import type { Request, Response } from 'express';

@Get()
public hello(@Req() req: Request, @Res() res: Response): void {}
```

**Expected Result**

**Actual Result**

In the code, `Request` and `Response` are used only as type information.

![image](https://user-images.githubusercontent.com/1300172/144955421-f6f0bfcf-8e76-474e-b0b5-92a07f58d054.png)

![image](https://user-images.githubusercontent.com/1300172/144955561-01beb5bf-6e7f-4398-a2cf-46fc708c50ff.png)

**Additional Info**

```ts
// This code also causes the same problem.
import type { Redis } from 'ioredis';

class Test {
  constructor(@Inject(REDIS) private redis: Redis) {}
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.6.0` |
| `@typescript-eslint/parser`        | `5.6.0` |
| `TypeScript`                       | `4.5.2` |
| `ESLint`                           | `8.3.0` |
| `node`                             | `16.13.0` |

**Comment by bradzacher** [comment]
TS is technically correct here - but it's inconsistent in its behaviour and impossible for us to properly reproduce in the lint rule.

`emitDecoratorMetadata` *implicitly* uses values - for example if you import a class and use it, TS will emit code which uses the class at runtime, even though your code as written only uses it as a type. This is consistent for most things that can be assigned to a type annotation - TS will implicitly use the value at runtime.

However if you pass an actual type (an interface, for example), TS will not use it as a value, and will instead use it as a type - it will use the type to inform the codegen. For example an interface will cause TS to emit `Object`, but a function type will cause TS to emit `Function`. TS does this across module boundaries (violating its own isolated modules build restrictions).

This is one of the few, rare cases where the two will disagree. Unfortunately because you need type information to do this, we cannot exactly replicate this behaviour in our lint rule. See other discussion on #3108.

You should either use the TS option, or our lint rule - not both.
